openhands/core/agent/base.py

```diff
@@ -67,6 +69,54 @@ def env_context(self) -> EnvContext | None:
         """Returns the environment context used by the Agent."""
         return self._env_context
 
+    @property
+    def confirmation_mode(self) -> bool:
+        """Returns whether confirmation mode is enabled."""
+        return self._confirmation_mode
+
+    def set_confirmation_mode(self, enabled: bool) -> None:
+        """Enable or disable confirmation mode."""
+        self._confirmation_mode = enabled
+
+    def get_pending_actions(self, state=None) -> list:
+        """Get list of pending actions awaiting confirmation.
+
+        Args:
+            state: ConversationState to scan for unmatched actions.
+                   If None, returns empty list (for backward compatibility).
+        """
+        if state is None:
+            return []
+        return self._get_unmatched_actions(state.events)
+
+    def clear_pending_actions(self) -> None:
+        """Clear all pending actions.
+
+        Note: This is a no-op in the history-based implementation since
+        pending actions are determined by scanning the event history.
+        Kept for backward compatibility.
+        """
+        pass
+
+    def _get_unmatched_actions(self, events) -> list:
+        """Find actions in the event history that don't have matching observations."""
+        from openhands.core.event import ActionEvent, ObservationEvent
+        from openhands.core.event.llm_convertible import UserRejectsObservation
+
+        # Get all action IDs that have observations
+        observed_action_ids = set()
+        for event in events:
+            if isinstance(event, (ObservationEvent, UserRejectsObservation)):
+                observed_action_ids.add(event.action_id)
+
+        # Find actions without matching observations
+        unmatched_actions = []
+        for event in events:
+            if isinstance(event, ActionEvent) and event.id not in observed_action_ids:
+                unmatched_actions.append(event)
+
+        return unmatched_actions
```

Comment:
I think we might want to save these to conversation state as we want the `Agent` to be as stateless as possible (i.e. not maintain its own state, so we only need to serialize the conversation for the agent to work)